engine/lib/analyses/mutable_variables.ml

```diff
@@ -0,0 +1,226 @@
+           (*         | _ -> m#zero) *)
+           (*        (Option.value_map subpat ~default:m#zero *)
+           (*           ~f:(fst >> super#visit_pat env))) *)
+           (*     (Option.value_map (Map.find env var) ~default:m#zero ~f:(fun x -> (Set.of_list (module W) x, Map.empty (module LocalIdent)))) *)
+
+           method visit_Let env _monadic pat expr body =
+             let new_set, new_env = super#visit_expr env expr in
+             m#plus
+               (super#visit_expr
+                  (Map.merge_skewed
+                     ~combine:(fun ~key:_ a b -> a @ b)
+                     (Map.merge_skewed
+                        ~combine:(fun ~key:_ a b -> a @ b)
+                        env new_env)
+                     (Map.of_alist_exn
+                        (module Local_ident)
+                        (List.map
+                           ~f:(fun v -> (v, Set.to_list new_set))
+                           (Set.to_list (U.Reducers.variables_of_pat pat)))))
+                  body)
+               (new_set, Map.empty (module Local_ident))
+
+           method! visit_local_ident (env : W.t list Map.M(Local_ident).t) ident
+               =
+             Option.value_map (Map.find env ident) ~default:m#zero ~f:(fun x ->
+                 (Set.of_list (module W) x, Map.empty (module Local_ident)))
+
+           method! visit_global_ident (_env : W.t list Map.M(Local_ident).t)
+               (x : Global_ident.t) =
+             match x with
+             | `Projector (`Concrete cid) | `Concrete cid -> (
+                 match
+                   Map.find data
+                     (Uprint.Concrete_ident_view.to_definition_name cid)
+                 with
+                 | Some (x, _) ->
+                     ( Set.of_list
+                         (module W)
+                         (List.map ~f:(fun x -> W.Identifier x) x),
+                       Map.empty (module Local_ident) )
+                 | _ -> m#zero)
+             | _ -> m#zero
+        end)
+          #visit_expr
+          env expr
+      in
+      (Set.to_list mut_var_set, new_env)
+  end
+
+  let rec analyse (func_dep : pre_data) (items : A.item list) : analysis_data =
+    let (mut_var_list, _)
+          : (string * (U.TypedLocalIdent.t * id_order) list) list * _ =
+      List.fold_left
+        ~f:(fun (y, count) x ->
+          match x.v with
+          | Fn { name; generics = _; body; params = _ } ->
+              let items, count = analyse_function_body body count in
+              ( y
+                @ [
+                    (Uprint.Concrete_ident_view.to_definition_name name, items);
+                  ],
+                count )
+          | Impl
+              {
+                generics = _;
+                self_ty = _;
+                of_trait = _ (* name, gen_vals *);
+                items;
+              } ->
+              List.fold_left
+                ~f:(fun (z, count) w ->
+                  match w.ii_v with
```

Comment:
This looks very similar to the code in `function_dependencies` line ~46.
We could have a `function_of_item: item -> (concrete_ident * expr) list option` that returns `Some (id, body)` on items that are functions with concrete identifier `id` and body `body`.
That would factorize things, and this function could live in `ast_utils`.